include/seqan3/alignment/matrix/debug_matrix.hpp

```diff
@@ -465,4 +483,14 @@ inline debug_stream_type<char_t> & operator<<(debug_stream_type<char_t> & s, ali
     return s;
 }
 
+//!\overload
+template <std::ranges::InputRange alignment_matrix_t, typename char_t>
+//!\cond
+    requires detail::DebugStreamableRange<alignment_matrix_t> && detail::Matrix<alignment_matrix_t>
+//!\endcond
+inline debug_stream_type<char_t> & operator<<(debug_stream_type<char_t> & s, alignment_matrix_t && matrix)
```

Comment:
Well, it is a long time ago and I thought it was part of another PR. But obviously not. This overload is required since my changes will add a range interface to the matrix. This however, competes with the overload for ranges of our debug stream. Accordingly a matrix, that is a range must refine the corresponding range overload. In the future, we can discuss if not the entire matrix concept should be adapted to refine the range concept as a two-dimensional range.  But I wasn't sure yet if we want this. And b) the required changes are beyond this PR.

Reply:
> This solution will not work on a standard compliant concepts implementation, because subsumption will only work on actual concept definitions (this changed in the standard at some point of time and is now wrong behaviour from gcc). That means that this overload is not more specialized than the Ranges one. We would need a new concept, e.g. RangableMatrix, RangeMatrix, or so, that refines Range.

So you saying:
```c++
template <typename t>
    requires ConceptA<t>
void foo();

template <typename t>
    requires ConceptA<t> && ConceptB<t>
void foo();
```

does not work but:


```c++
template <typename t>
    requires ConceptA<t>
void foo();

template <typename t>
concept ConceptC = ConceptA<t> && ConceptB<t>;

template <typename t>
    requires ConceptC<t>
void foo();
```
does work? What am I missing here?
Note that in the above code everything is a concept defintion and that the ranges overload uses the first two concepts in the exact same order.

```cpp
template <std::ranges::InputRange t, typename char_t>
    requires detail::DebugStreamableRange<t>
inline debug_stream_type<char_t> & operator<<(...)
{...}
```
So just aliasing the expression as with another concept name should not make a difference, should it?

Reply:
>I imagine that the range of the matrix is over all elements? Or is it a range over the columns? Or is it a range over the row? Because of this ambiguity, I would not make the matrix a range itself, but add functions like columns(), rows(), entries() to make it clear (those give then ranges back).

It implements a simple range interface providing a two_dimensional iterator. the layout of the matrix is independent of that. You always iterate along the major column order of the matrix. There is no ambiguity.

Reply:
>In general this use-case might give us headaches in other areas too. If one type satisfies two distinct concepts, but there are two overloads for the two concepts, they will fail on you. I have an idea how to find this problems with CI.
We should have a new test suite (like /test/unit, test/coverage) where every test header includes the complete library (seqan3/all.hpp) (normally a test only includes a subset and might miss this problem). By doing this we will get compiler errors for this use case and can fix them.

Yes, we already discussed separating unit tests and coverage tests. If you have an idea to improve it please add a card on the test module board. But for now this has absolutely no priority!!!

Reply:
> > This solution will not work on a standard compliant concepts implementation, because subsumption will only work on actual concept definitions (this changed in the standard at some point of time and is now wrong behaviour from gcc). That means that this overload is not more specialized than the Ranges one. We would need a new concept, e.g. RangableMatrix, RangeMatrix, or so, that refines Range.
> 
> So you saying:
> 
> ```c++
> template <typename t>
>     requires ConceptA<t>
> void foo();
> 
> template <typename t>
>     requires ConceptA<t> && ConceptB<t>
> void foo();
> ```
> 
> does not work but:
> 
> ```c++
> template <typename t>
>     requires ConceptA<t>
> void foo();
> 
> template <typename t>
> concept ConceptC = ConceptA<t> && ConceptB<t>;
> 
> template <typename t>
>     requires ConceptC<t>
> void foo();
> ```
> 
> does work? What am I missing here?

Yes this is working! That's exactly what I wanted to express. 

See https://en.cppreference.com/w/cpp/language/constraints#Atomic_constraints they give an example:

```c++
template<class T> constexpr bool is_meowable = true;
template<class T> constexpr bool is_cat = true;
 
template<class T>
concept Meowable = is_meowable<T>;
 
template<class T>
concept BadMeowableCat = is_meowable<T> && is_cat<T>;
 
template<class T>
concept GoodMeowableCat = Meowable<T> && is_cat<T>;
 
template<Meowable T>
void f1(T); // #1
 
template<BadMeowableCat T>
void f1(T); // #2
 
template<Meowable T>
void f2(T); // #3
 
template<GoodMeowableCat T>
void f2(T); // #4
 
void g(){
    f1(0); // error, ambiguous:
           // the is_meowable<T> in Meowable and BadMeowableCat forms distinct
           // atomic constraints that are not identical (and so do not subsume each other)
 
    f2(0); // OK, calls #4, more constrained than #3
           // GoodMeowableCat got its is_meowable<T> from Meowable
}
```

> Note that in the above code everything is a concept defintion and that the ranges overload uses the first two concepts in the exact same order.
> 
> ```c++
> template <std::ranges::InputRange t, typename char_t>
>     requires detail::DebugStreamableRange<t>
> inline debug_stream_type<char_t> & operator<<(...)
> {...}
> ```
> 
> So just aliasing the expression as with another concept name should not make a difference, should it?

It does.

Reply:
Mhmm, I still have issues with this. According to partial ordering of constraints: `If declarations D1 and D2 are constrained and D1's associated constraints subsume D2's associated constraints (or if D2 is unconstrained), then D1 is said to be at least as constrained as D2. If D1 is at least as constrained as D2 and D2 is not at least as constrained as D1, then D1 is more constrained than D2. `

So I cannot read from this, that a conjunction of atomic constraints subsumes only another conjunction of constraints if it was first aliased into another constrained. 
Why should `A & B & C` not apply to these rules and `D = A & B & C` does? Can we discuss this offline?

Reply:
Ok, I tested it on Godbolt and I have to say that I am right 🏆 🤣 
No, joke aside: https://godbolt.org/z/rMGnpn 
Please help me understand what I am not seeing here. The example as given does work with the current clang concepts branch and is a simplified form of the version above.